src/node_buffer.cc

```diff
@@ -855,87 +856,149 @@ void Compare(const FunctionCallbackInfo<Value> &args) {
 }
 
 
-int32_t IndexOf(const char* haystack,
-                size_t h_length,
-                const char* needle,
-                size_t n_length) {
-  CHECK_GE(h_length, n_length);
-  // TODO(trevnorris): Implement Boyer-Moore string search algorithm.
-  for (size_t i = 0; i < h_length - n_length + 1; i++) {
-    if (haystack[i] == needle[0]) {
-      if (memcmp(haystack + i, needle, n_length) == 0)
-        return i;
-    }
-  }
-  return -1;
-}
-
-
 void IndexOfString(const FunctionCallbackInfo<Value>& args) {
   ASSERT(args[1]->IsString());
   ASSERT(args[2]->IsNumber());
 
+  enum encoding enc =
+      args[3].IsEmpty() ? UTF8 : ParseEncoding(args.GetIsolate(),
+                                               args[3].As<String>(),
+                                               UTF8);
```

Comment:
No need to short-circuit. `ParseEncoding()` already does, and I doubt there'd be enough performance gain to do it one call earlier.